### ISSUE TITLE
gpu-api=vulkan causes issues on video switch

### ISSUE THREAD

**Issue by LisandraBrave** [mpv-player/mpv#7923]
`$ mpv --version`
mpv 0.32.0 Copyright © 2000-2020 mpv/MPlayer/mplayer2 projects
 built on UNKNOWN
ffmpeg library versions:
   libavutil       56.31.100
   libavcodec      58.54.100
   libavformat     58.29.100
   libswscale      5.5.100
   libavfilter     7.57.100
   libswresample   3.5.100
ffmpeg version: 4.2.4

Fedora 32 kernel ver 5.7.8-200.fc32.x86_64
MPV installed from `rpmfusion-free-updates`
`$ openbox --version`
Openbox 3.6.1
NVIDIA GeForce RTX 2060, running on NVIDIA driver 450.57

### Reproduction steps

1. Load multiple files in a playlist
2. wait for one video to finish playing

Does not occur when `gpu-api=vulkan` is not set in mpv.conf. No behavior change whether or not `profile=gpu-hq` is in configuration.

### Expected behavior
Next video in playlist plays

### Actual behavior
Entire video output hangs for a few seconds, then window no longer renders anything, while sound continues to play. `dmesg -w` logs `NVRM: Xid (PCI:0000:07:00): 31, pid=35313, Ch 0000002b, intr 00000000. MMU Fault: ENGINE GRAPHICS GPCCLIENT_PROP_0 faulted @ 0x1_2df47000. Fault is of type FAULT_PDE ACCESS_TYPE_VIRT_WRITE`

[Log file](https://github.com/mpv-player/mpv/files/4929525/output.txt)

**Comment by haasn** [comment]
Can you post a log with `--gpu-debug`?

**Comment by LisandraBrave** [comment]
[output.txt](https://github.com/mpv-player/mpv/files/4929738/output.txt)

**Comment by haasn** [comment]
It seems like you do not have the vulkan debug layers installed, for `--gpu-debug` to work. Try looking for a package like "vulkan-layers" on your distro.

**Comment by LisandraBrave** [comment]
Bizzarely, installing that appears to have solved the initial issue on video switch, but now it crashes when I attempt to resize the window.
[output.txt](https://github.com/mpv-player/mpv/files/4929888/output.txt)

**Comment by haasn** [comment]
Judging by the error messages it seems to be some sort of destroy-while-still-in-use, although I'm not sure what could cause it because we refcount all textures. Do you think you could use `gdb` to find the offending `pl_tex_destroy` and figure out which texture is triggering the validation errors?

It would probably help if you could upgrade to the latest version of `libplacebo`, ideally git master. (That might also fix the issue!)

**Comment by LisandraBrave** [comment]
[placebo.txt](https://github.com/mpv-player/mpv/files/4930244/placebo.txt)
I have no idea how to find which texture is causing the issue. This is gdb stepping through. Now to figure out how to upgrade libplacebo on my system.

**Comment by haasn** [comment]
Ah, `29`. That's a rather old version of libplacebo. It absolutely would not surprise me if this was fixed on master. When you step through the code can you use `bt` to get a backtrace of whereabouts this call is coming from? (Line number information would also help, although I realize this may be non-trivial to set up on fedora)

**Comment by LisandraBrave** [comment]
[placebo.txt](https://github.com/mpv-player/mpv/files/4930675/placebo.txt)
This is `step`ping through, with a `bt` at the first opportunity after the `VK_ERROR_DEVICE_LOST`. Strangely, it went to the next playlist entry right after it threw the error. Also strange is that there should be nothing activating the screensaver, since xscreensaver is set with `lock: False`.

**Comment by Nocifer** [comment]
Hello,

I've been experiencing the same issue on Arch Linux with an Nvidia 1660. I think I've narrowed it down to having `fullscreen` enabled on launch or not. Whatever my other settings, if I open a video with `fullscreen=yes` either in my mpv.conf or in its watch_later file, then if I quit mpv and open the video again for a second or third time (or even open a different video, it makes no difference, it just needs to be launched in fullscreen) the issue can consistently be reproduced. It takes 2 to 3 or at most 4 launches.

I've experimented both with and without HW decoding enabled, both with the beta Nvidia vulkan drivers and with the stable ones, both with libplacebo 1.29 from the standard repos and with a self-compiled 2.72+, all in vain. The one thing that seems to make a difference is `fullscreen`.

Here's an example [output.log](https://github.com/mpv-player/mpv/files/5052305/output.log) with `gpu-debug` enabled. From line 2749 onwards it just keeps spamming that last error ad nauseam until I force kill mpv, so I've removed the rest of the log.

**Comment by philipl** [comment]
Where is your sharderc/glslang coming from? I had issues with the lunarg builds last week as they went through a major change in their build system (switching to Ubuntu 20.04) and I saw this error during that time. It might have been a build problem or a bad snapshot, which happens with those projects.

**Comment by Nocifer** [comment]
Both are installed from the standard repos, so I presume they're as vanilla as it gets. Though I wouldn't exclude the possibility that this is an anomaly like you describe which affects even the official builds, it would not be the first time something like that has happened. Maybe that's the reason why I/we only noticed this issue now.

**Comment by Nocifer** [comment]
Alright, this is after all kind of a false alarm, at least as far as mpv is concerned. While messing around with another program that utilizes Vulkan (RPCS3, an emulator for PS3) and while everything worked flawlessly at first, as soon as I set it to launch the games directly in fullscreen, I got a suspiciously similar error in the log and the games began hanging and crashing just like mpv.

IIRC the error included some "vk::wait" call that timed out and a warning that the GPU device couldn't be found; unfortunately I didn't copy the error from the log immediately and now the log has been already overwritten, while the error itself cannot be reproduced anymore, at least for the time being.

Still, I now suspect that this is some kind of shenanigan not of mpv but either the Nvidia driver, GNOME's Mutter, or Vulkan itself.

**Comment by 115100** [comment]
I'm not longer getting this after upgrading to nvidia driver 450.66.

**Comment by Nocifer** [comment]
> I'm no longer getting this after upgrading to nvidia driver 450.66.

Yeah, this tidbit from the new driver's changelog

`"Fixed triple buffering support of Vulkan X11 swapchains when applications are syncing to vblank."`

might have something to do with that. I'm still on the beta 450.56.06 driver, so I can't test it myself, but I'll report back when I do eventually update to a driver version that includes this fix.

**Comment by Nocifer** [comment]
Upgrading to the new Nvidia vulkan 455.24.02 driver, which also includes all the fixes introduced in the beta 450.66 driver, has fixed the issue for me in 99.9% of cases. Out of some ~50 or so times I tried to force the issue to manifest, I only managed to do it once, and that one time may have something to do with me accidentally opening two fullscreen instances of mpv at the same time, which may have caused the compositor to go bonkers. Unfortunately I didn't have debug logging enabled when it happened.

But anyway, I guess this issue can now be safely closed, especially since there's already that older report by @115100 that it's already been fixed.
